### PR TITLE
BDD: Added static ordering support and dynamic reordering for BDDs

### PR DESCRIPTION
**Pull request by mikand**
Added support for dynamic reordering and specification of an initial ordering for BddSolver.

This addresses #89

### CODE REVIEW

File: pysmt/solvers/bdd.py [hunk 1]
```diff
@@ -54,13 +55,56 @@ def __enter__(self):
...
+            LockDdManager._depth -= 1
+            if LockDdManager._depth == 0:
+                pycudd.ResetDefaultDdManager()
+
+
+
+class BddOptions(SolverOptions):
+
+    ## CUDD Reordering algorithms
+    CUDD_REORDER_SAME, \
```
Comment (marcogario) [inline pysmt/solvers/bdd.py]: Indentation.

File: pysmt/solvers/bdd.py [hunk 2]
```diff
@@ -54,13 +55,56 @@ def __enter__(self):
...
+        CUDD_REORDER_WINDOW3_CONV, \
+        CUDD_REORDER_WINDOW4_CONV, \
+        CUDD_REORDER_GROUP_SIFT, \
+        CUDD_REORDER_GROUP_SIFT_CONV, \
+        CUDD_REORDER_ANNEALING, \
+        CUDD_REORDER_GENETIC, \
+        CUDD_REORDER_LINEAR, \
+        CUDD_REORDER_LINEAR_CONVERGE, \
+        CUDD_REORDER_LAZY_SIFT, \
+        CUDD_REORDER_EXACT = range(22)
```
Comment (marcogario) [inline pysmt/solvers/bdd.py]: This is not defined anywhere by pycudd?

File: pysmt/solvers/bdd.py [hunk 3]
```diff
@@ -87,6 +148,22 @@ def __init__(self, environment, logic, user_options):
         self.backtrack = []
         self.latest_model = None
 
+    def get_default_options(self, logic=None, user_options=None):
+        res = BddOptions()
+        if user_options is not None:
+            if "unsat_cores_mode" in user_options:
```
Comment (marcogario) [inline pysmt/solvers/bdd.py]: So, the proper way to create this is {"unsat_cores_mode": None}? Is this because of how the factory works? Wouldn't it be better just to assert that "unsat_cores_mode" not in user_options?

File: pysmt/solvers/bdd.py [hunk 4]
```diff
@@ -87,6 +148,22 @@ def __init__(self, environment, logic, user_options):
...
+                if user_options["unsat_cores_mode"] != None:
+                    raise NotImplementedError("BddSolver does not "\
+                                              "support unsat cores")
+
+            for k in ["static_ordering",
```
Comment (marcogario) [inline pysmt/solvers/bdd.py]: Maybe you can post-pone the initialization of BddOptions till here and do the unpacking of keywords arguments? If they are not defined, you will take the default
```
 BddOptions(**user_options)
```
I think it is nicer and less error-prone than *setattr*

File: pysmt/test/test_bdd.py [hunk 5]
```diff
@@ -115,6 +122,41 @@ def test_quantifier_eliminator(self):
         g = qelim(f, solver_name="bdd")
         self.assertEqual(g, self.y)
 
+    @skipIfSolverNotAvailable("bdd")
+    def test_reordering(self):
+        from pysmt.solvers.bdd import BddSolver
```
Comment (marcogario) [inline pysmt/test/test_bdd.py]: We should extend the factory to be able to pass user_options to the solver. Thus this would simply be a call to the factory *Solver* .

File: pysmt/test/test_bdd.py [hunk 6]
```diff
@@ -115,6 +122,41 @@ def test_quantifier_eliminator(self):
...
+            self.assertTrue(s.solve())
+
+    @skipIfSolverNotAvailable("bdd")
+    def test_reordering_algorithms(self):
+        from pysmt.solvers.bdd import BddSolver, BddOptions
+        for algo in BddOptions.CUDD_ALL_REORDERING_ALGORITHMS:
+            with BddSolver(get_env(), pysmt.logics.BOOL,
+                           {"dynamic_reordering" : True,
+                            "reordering_algorithm" : algo}) as s:
+                s.add_assertion(self.big_tree)
```
Comment (marcogario) [inline pysmt/test/test_bdd.py]: Maybe query the ddmanager and check that the current reordering is the one specified.

File: pysmt/test/test_bdd.py [hunk 7]
```diff
@@ -115,6 +122,41 @@ def test_quantifier_eliminator(self):
...
+                self.assertTrue(s.solve())
+
+    @skipIfSolverNotAvailable("bdd")
+    def test_fixed_ordering(self):
+        from pysmt.solvers.bdd import BddSolver
+        with BddSolver(get_env(), pysmt.logics.BOOL,
+                       {"static_ordering" : [self.x, self.y]}) as s:
+            s.add_assertion(self.big_tree)
```
Comment (marcogario) [inline pysmt/test/test_bdd.py]: Add case in which ordering is invalid (i.e., non-bool, or strings).

File: pysmt/test/test_bdd.py [hunk 8]
```diff
@@ -115,6 +122,41 @@ def test_quantifier_eliminator(self):
...
+            self.assertTrue(s.solve())
+
+    @skipIfSolverNotAvailable("bdd")
+    def test_initial_ordering(self):
+        from pysmt.solvers.bdd import BddSolver
+        with BddSolver(get_env(), pysmt.logics.BOOL,
+                       {"static_ordering" : [self.x, self.y],
+                        "dynamic_reordering" : True}) as s:
+            s.add_assertion(self.big_tree)
```
Comment (marcogario) [inline pysmt/test/test_bdd.py]: How can we test that this is really working? This test needs to better capture the semantic of this operation.

File: pysmt/solvers/bdd.py (hunk 2 above)
Comment (marcogario) [inline pysmt/solvers/bdd.py]: no, seems like no. This is a bit fragile, but until we have a better wrapper for cudd, this is ok.

File: pysmt/test/test_bdd.py (hunk 5 above)
Comment (mikand) [inline pysmt/test/test_bdd.py]: This is an orthogonal issue. See #105
